### ISSUE TITLE
Inconsistent code coverage in non-executed classes

### ISSUE THREAD

**Issue by CJDennis** [Codeception/Codeception#5768]
#### What are you trying to achieve?
Consistent code coverage. 

#### What do you get instead?
Inconsistent code coverage. Lines 10, 11, & 13 are correctly marked as not executable in `CJDennis\ElseTest\IfTest`, while they are incorrectly marked as executable and not covered in `CJDennis\ElseTest\ElseTest`. Running the same coverage report under PHPUnit gives consistent results (neither file has those lines marked as executable).

> Provide console output if related. Use `-vvv` mode for more details.

```bash
codecept run unit --coverage-xml coverage.xml
```

### tests/_output/coverage.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<coverage generated="1574649066">
  <project timestamp="1574649066">
    <package name="CJDennis\ElseTest">
      <file name="D:\server\Apache24\htdocs\libraries\cjdennis\else-test\src\CJDennis\ElseTest\ElseTest.php">
        <class name="CJDennis\ElseTest\ElseTest" namespace="CJDennis\ElseTest">
          <metrics complexity="2" methods="1" coveredmethods="0" conditionals="0" coveredconditionals="0" statements="7" coveredstatements="0" elements="8" coveredelements="0"/>
        </class>
        <line num="7" type="method" name="__construct" visibility="public" complexity="2" crap="6" count="0"/>
        <line num="8" type="stmt" count="0"/>
        <line num="9" type="stmt" count="0"/>
        <line num="10" type="stmt" count="0"/>
        <line num="11" type="stmt" count="0"/>
        <line num="12" type="stmt" count="0"/>
        <line num="13" type="stmt" count="0"/>
        <line num="14" type="stmt" count="0"/>
        <metrics loc="15" ncloc="15" classes="1" methods="1" coveredmethods="0" conditionals="0" coveredconditionals="0" statements="7" coveredstatements="0" elements="8" coveredelements="0"/>
      </file>
      <file name="D:\server\Apache24\htdocs\libraries\cjdennis\else-test\src\CJDennis\ElseTest\IfTest.php">
        <class name="CJDennis\ElseTest\IfTest" namespace="CJDennis\ElseTest">
          <metrics complexity="2" methods="1" coveredmethods="0" conditionals="0" coveredconditionals="0" statements="4" coveredstatements="3" elements="5" coveredelements="3"/>
        </class>
        <line num="7" type="method" name="__construct" visibility="public" complexity="2" crap="2.06" count="1"/>
        <line num="8" type="stmt" count="1"/>
        <line num="9" type="stmt" count="0"/>
        <line num="12" type="stmt" count="1"/>
        <line num="14" type="stmt" count="1"/>
        <metrics loc="15" ncloc="15" classes="1" methods="1" coveredmethods="0" conditionals="0" coveredconditionals="0" statements="4" coveredstatements="3" elements="5" coveredelements="3"/>
      </file>
    </package>
    <metrics files="2" loc="30" ncloc="30" classes="2" methods="2" coveredmethods="0" conditionals="0" coveredconditionals="0" statements="11" coveredstatements="3" elements="13" coveredelements="3"/>
  </project>
</coverage>
```

> Provide test source code if related

### tests/unit/CJDennis/ElseTest/IfTestTest.php
```php
<?php
namespace CJDennis\ElseTest;

use Codeception\Test\Unit;
use UnitTester;

class IfTestTest extends Unit {
  /** @var UnitTester */
  protected $tester;

  protected function _before() {
  }

  protected function _after() {
  }

  // tests
  public function testSomeFeature() {
    new IfTest();
  }
}
```

### src/CJDennis/ElseTest/IfTest.php
```php
<?php
namespace CJDennis\ElseTest;

class IfTest {
  protected $test;

  public function __construct() {
    if (false) {
      $this->test = true;
    }
    else {
      $this->test = false;
    }
  }
}
```

### src/CJDennis/ElseTest/ElseTest.php
```php
<?php
namespace CJDennis\ElseTest;

class ElseTest {
  protected $test;

  public function __construct() {
    if (false) {
      $this->test = true;
    }
    else {
      $this->test = false;
    }
  }
}
```
### Details

* Codeception version: 3.1.2
* PHP Version: 7.3.9
* Operating System: Windows
* Installation type: Composer
* List of installed packages (`composer show`)
```
behat/gherkin                      v4.6.0     Gherkin DSL parser for PHP 5.3
codeception/codeception            3.1.2      BDD-style testing framework
codeception/phpunit-wrapper        8.1.0      PHPUnit classes used by Codeception
codeception/stub                   3.6.0      Flexible Stub wrapper for PHPUnit's Mock Builder
doctrine/instantiator              1.3.0      A small, lightweight utility to instantiate objects in PHP without invoking their constructors
facebook/webdriver                 1.7.1      A PHP client for Selenium WebDriver
guzzlehttp/guzzle                  6.4.1      Guzzle is a PHP HTTP client library
guzzlehttp/promises                v1.3.1     Guzzle promises library
guzzlehttp/psr7                    1.6.1      PSR-7 message implementation that also provides common utility methods
hoa/consistency                    1.17.05.02 The Hoa\Consistency library.
hoa/console                        3.17.05.02 The Hoa\Console library.
hoa/event                          1.17.01.13 The Hoa\Event library.
hoa/exception                      1.17.01.16 The Hoa\Exception library.
hoa/file                           1.17.07.11 The Hoa\File library.
hoa/iterator                       2.17.01.10 The Hoa\Iterator library.
hoa/protocol                       1.17.01.14 The Hoa\Protocol library.
hoa/stream                         1.17.02.21 The Hoa\Stream library.
hoa/ustring                        4.17.01.16 The Hoa\Ustring library.
myclabs/deep-copy                  1.9.3      Create deep copies (clones) of your objects
phar-io/manifest                   1.0.3      Component for reading phar.io manifest information from a PHP Archive (PHAR)
phar-io/version                    2.0.1      Library for handling version information and constraints
phpdocumentor/reflection-common    2.0.0      Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock  4.3.2      With this component, a library can provide support for annotations via DocBlocks or otherwise retrieve information that is embedded in a DocBlock.
phpdocumentor/type-resolver        1.0.1      A PSR-5 based resolver of Class names, Types and Structural Element Names
phpspec/prophecy                   1.9.0      Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage          7.0.10     Library that provides collection, processing, and rendering functionality for PHP code coverage information.
phpunit/php-file-iterator          2.0.2      FilterIterator implementation that filters files based on a list of suffixes.
phpunit/php-text-template          1.2.1      Simple template engine.
phpunit/php-timer                  2.1.2      Utility class for timing
phpunit/php-token-stream           3.1.1      Wrapper around PHP's tokenizer extension.
phpunit/phpunit                    8.4.3      The PHP Unit Testing framework.
psr/container                      1.0.0      Common Container Interface (PHP FIG PSR-11)
psr/http-message                   1.0.1      Common interface for HTTP messages
ralouphie/getallheaders            3.0.3      A polyfill for getallheaders.
sebastian/code-unit-reverse-lookup 1.0.1      Looks up which function or method a line of code belongs to
sebastian/comparator               3.0.2      Provides the functionality to compare PHP values for equality
sebastian/diff                     3.0.2      Diff implementation
sebastian/environment              4.2.3      Provides functionality to handle HHVM/PHP environments
sebastian/exporter                 3.1.2      Provides the functionality to export PHP variables for visualization
sebastian/global-state             3.0.0      Snapshotting of global state
sebastian/object-enumerator        3.0.3      Traverses array structures and object graphs to enumerate all referenced objects
sebastian/object-reflector         1.1.1      Allows reflection of object attributes, including inherited and non-public ones
sebastian/recursion-context        3.0.0      Provides functionality to recursively process PHP variables
sebastian/resource-operations      2.0.1      Provides a list of PHP built-in functions that operate on resources
sebastian/type                     1.1.3      Collection of value objects that represent the types of the PHP type system
sebastian/version                  2.0.1      Library that helps with managing the version number of Git-hosted PHP projects
symfony/browser-kit                v4.4.0     Symfony BrowserKit Component
symfony/console                    v4.4.0     Symfony Console Component
symfony/css-selector               v4.4.0     Symfony CssSelector Component
symfony/dom-crawler                v4.4.0     Symfony DomCrawler Component
symfony/event-dispatcher           v4.4.0     Symfony EventDispatcher Component
symfony/event-dispatcher-contracts v1.1.7     Generic abstractions related to dispatching event
symfony/finder                     v4.4.0     Symfony Finder Component
symfony/polyfill-ctype             v1.12.0    Symfony polyfill for ctype functions
symfony/polyfill-mbstring          v1.12.0    Symfony polyfill for the Mbstring extension
symfony/polyfill-php73             v1.12.0    Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/process                    v4.4.0     Symfony Process Component
symfony/service-contracts          v2.0.0     Generic abstractions related to writing services
symfony/yaml                       v4.4.0     Symfony Yaml Component
theseer/tokenizer                  1.1.3      A small library for converting tokenized PHP source code into XML and potentially other formats
webmozart/assert                   1.6.0      Assertions to validate method input/output with nice error messages.
```
* Suite configuration:

### codeception.yml
```yml
paths:
    tests: tests
    output: tests/_output
    data: tests/_data
    support: tests/_support
    envs: tests/_envs
actor_suffix: Tester
extensions:
    enabled:
        - Codeception\Extension\RunFailed
coverage:
    enabled: true
    include:
        - src/*
    low_limit: 50
    high_limit: 100
```

### tests/unit.suite.yml
```yml
# Codeception Test Suite Configuration
#
# Suite for unit or integration tests.

actor: UnitTester
modules:
    enabled:
        - Asserts
        - \Helper\Unit
```

**Comment by Naktibalda** [comment]
This is not an issue of Codeception, it is an issue of xdebug or pcov (if you use it as an alternative to xdebug).
Do you use the same version of xdebug, PHPUnit and `phpunit/php-code-coverage` when you run those tests with Codeception and when you run it with PHPUnit?

**Comment by CJDennis** [comment]
This is my `composer.json`

```json
{
    "name": "cjdennis/else-test",
    "require-dev": {
        "codeception/codeception": "^3.1"
    },
    "license": "MIT",
    "autoload": {
        "psr-4": {"CJDennis\\ElseTest\\": "src/CJDennis/ElseTest/"}
    },
    "require": {}
}
```

Both sets of tests are using PHPUnit 8.4.3. PHP is configured the same.

**Comment by dekker-m** [comment]
I actually don't think this is related to xdebug or pcov. I've ran into a similar issue with code coverage yesterday (when I switched from xdebug to pcov). From what I can see from the results I guess you are using pcov instead a xdebug as well. (Could you post the output of `php -i`, and give us the Xdebug and Pcov settings if they are present in the output ?)

I did some debugging today to figure out how to explain the difference. Xdebug seems to merge the results correctly, while pcov does not. At first I looked at the data returned by the driver classes from the php-code-coverage package. Both drivers seem to return similar results on individual test cases. xdebug marks some lines as "dead code", while pcov marks them as "not executable". Those small differences could not explain the output you received as well. In short, it is not related to xdebug or pcov at all.

So far, both drivers seem to return the same results, right? I was still wondering why the final result did not yield the same result as the individual tests. Codeception uses php-code-coverage its CodeCoverage::merge() method to merge results from different tests/suits. Finally, before everything is printed to your screen or dumped to an XML file, Codeception merges the coverage results again. Somewhere in those final steps the additional lines of not executed code are added.

It all seemed to be related to one of PHPUnit's settings: `addUncoveredFilesFromWhitelist`. Before php-code-coverage returns the results, every file on the whitelist is read from disk and their lines are added to the result as not executed. The settings of "show_uncovered: true/false` in your codeception.yml do not seem to matter. I made some local changes to codeception:

`NonFinal/TestRunner.php (line 1214, phpunit-wrapper package)`
```
$arguments['addUncoveredFilesFromWhitelist']                  = $arguments['addUncoveredFilesFromWhitelist'] ?? false;
```

`Codeception\Coverage\SuiteSubscriber (add line to applySettings)`
```
    protected function applySettings($settings)
    {
        /// ...
        $this->coverage->setAddUncoveredFilesFromWhitelist(false);
    }
```

`Codeception\Coverage\Subscriber\Printer (added line to __construct)`
```
    public function __construct($options)
    {
        // ...

        self::$coverage = new \SebastianBergmann\CodeCoverage\CodeCoverage();
        self::$coverage->setAddUncoveredFilesFromWhitelist(false);

        // ...
    }
```

After those changes the code coverage results seemed to be consistent between Xdebug/pcov and phpunit/codeception.

@Naktibalda I think the changes to the files above should fix a lot of code coverage issues from pcov in codeception (and maybe some from xdebug in codeception as well). The results with `$addUncoveredFilesFromWhitelist = true` still exist if `show_uncovered: true` is used and multiple suites are executed at the same time. Could you validate the solution provided above?

**Comment by Naktibalda** [comment]
@dekker-m `show_uncovered` setting must be set in `coverage` section:

```yaml
coverage:
    enabled: true
    show_uncovered: false
```

What's the difference between `setAddUncoveredFilesFromWhitelist` and `setProcessUncoveredFilesFromWhitelist`?

**Comment by dekker-m** [comment]
I've set `show_uncovered` to `true`, but it does not resolve the issue when running multiple suits or using Robo for parallel testing. They are all different scenario's of running codeception, but all issues seem to lead back to `addUncoveredFilesFromWhitelist`.

`addUncoveredFilesFromWhitelist` seems to add all uncovered files from the whitelist when you receive the data. As you can see in:

https://github.com/sebastianbergmann/php-code-coverage/blob/f1884187926fbb755a9aaf0b3836ad3165b478bf/src/CodeCoverage.php#L188-L198

This method is used by the report builder from the php-code-coverage library. The method which actually adds all data:

https://github.com/sebastianbergmann/php-code-coverage/blob/f1884187926fbb755a9aaf0b3836ad3165b478bf/src/CodeCoverage.php#L575-L606

As for codeception, when `show_uncovered`  is set to `true`, `processUncoveredFilesFromWhitelist` should have loaded all files already. `processUncoveredFilesFromWhitelist` actually works better because it uses xdebug or pcov to determine which lines are executable.  When `addUncoveredFilesFromWhitelist()` is called as well, it screws up pcov data.

When `show_uncovered` is set to `false`, it should not be called at all. I think the expected behavior is that uncovered files do not show up in the report. Because `addUncoveredFilesFromWhitelist` is true by default, all files from the whitelists are still included.

The difference between `processUncoveredFilesFromWhitelist` and `addUncoveredFilesFromWhitelist` is the way the lines are loaded.  The method `processUncoveredFilesFromWhitelist` loads all whitelisted files _before_ at initialization. It does so by using pcov or xdebug to determine executable code. On the other hand, `addUncoveredFilesFromWhitelist` adds the files _after_ the tests are executed. It does seem to add uncovered files from the whitelist, but without determining which lines can be executed and which lines cannot.

This would explain why `ElseTest` does not show executable code properly in @CJDennis his code (show_uncovered is not set in the configuration, show_uncovered's default value is false).

**Comment by Naktibalda** [comment]
It would be great if you could create a test in [tests/coverage](https://github.com/Codeception/Codeception/tree/4.0/tests/coverage) to demonstrate the issue and then fix it.

**Comment by dekker-m** [comment]
I tried to get the docker container running for the test cases, but so far no luck. The first error I ran into showed the requirements could not be resolved to an installable set of packages

```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Can only install one of: codeception/codeception[4.0.x-dev, No version set (parsed as 1.0.0)].
    - codeception/module-filesystem 1.0.1 conflicts with codeception/codeception[No version set (parsed as 1.0.0)].
    - codeception/module-filesystem 1.0.2 conflicts with codeception/codeception[No version set (parsed as 1.0.0)].
    - codeception/module-filesystem dev-master conflicts with codeception/codeception[No version set (parsed as 1.0.0)].
    - Installation request for codeception/codeception No version set (parsed as 1.0.0) -> satisfiable by codeception/codeception[No version set (parsed as 1.0.0)].
    - codeception/module-filesystem 1.0.0 requires codeception/codeception 4.0.x-dev -> satisfiable by codeception/codeception[4.0.x-dev].
    - Installation request for codeception/module-filesystem *@dev -> satisfiable by codeception/module-filesystem[1.0.0, 1.0.1, 1.0.2, dev-master].
```

I'm not sure if this is supposed to happen, but I added a `"version": "4.0.x-dev"` to composer.json to get composer to install the packages (I don't see this in the documentation, is it supposed to do this?).

After the packages are installed, it has some trouble generating the auto loader:

```
  [RuntimeException]
  Could not scan for classes inside "tests/cli/_steps" which does not appear to be a file nor a folder


install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--] [<packages>]...

Service 'codecept' failed to build: The command '/bin/sh -c composer install --prefer-dist --no-interaction --optimize-autoloader --apcu-autoloader' returned a non-zero code: 1
```

Do you have any idea why the docker environment isn't properly installed? I never worked with Docker before (we use Vagrant to run our dev environments). The Dockerfile seems to install composer before it adds the source of codeception to the container, is it supposed to work in this order?

I'll see if I can write a test case the demonstrate the behavior as soon as I get the docker container running. 

Btw, which branch should I use to create the test cases?

**Comment by Naktibalda** [comment]
Dockerfile wasn't updated for Codeception 4, don't use it.

**Comment by dekker-m** [comment]
I'm on Windows and my user folder got a space in it. Executing shell commands is very buggy because the spaces in path names are not properly escaped when codeception runs shell commands. I won't be able to run any coverage test if docker is not working. Anyway, with a few minor changes to the Dockerfile and composer.json I got it running.

Anyway, I created a new branch on my fork: https://github.com/Codeception/Codeception/compare/4.0...dekker-m:bugfix/show-uncovered

To summarize:
* math.suite.yml contained a typo
* I added a second math_show_uncovered.suite.yml file with the same settings as the original, but with show_uncovered set to true.
* The LocalShowUncoveredCept shows how I would expect show_uncovered to work. The test runs runs the following:
** MathCest::testAddition from the math_show_uncovered suite. It will only test addition, so the subtract class should not be covered. However, because show_uncovered is set to true, the total coverage should be 50%.
** MathCest::testAddition from the math suite. show_uncovered is not set and because the subtract class is never covered, it should not be taken into account when the results are calculated. The addition class is covered though. We should expect 100% code coverage here, but it does not because `addUncoveredFilesFromWhitelist` is still set to `true`.

Do you agree the test case shows the expected behavior?

**Comment by dekker-m** [comment]
I've update the branch in my fork. The results from the tests are:

```
$ docker-compose run --rm codecept run coverage LocalShowUncoveredCept
Codeception PHP Testing Framework v4.1.1
Powered by PHPUnit 9.0.1 by Sebastian Bergmann and contributors.
Running with seed:

[Seed] 2066400396

Coverage Tests (1) -----------------------------------------
✖ LocalShowUncoveredCept: Run local code coverage for cest and test without show uncovered (1.00s)
------------------------------------------------------------


Time: 3.61 seconds, Memory: 10.00 MB

There was 1 failure:

---------
1) LocalShowUncoveredCept: Run local code coverage for cest and test without show uncovered
 Test  tests/coverage/LocalShowUncoveredCept.php
 Step  See in shell output "Classes: 100.00%"
 Fail  Failed asserting that 'Codeception PHP Testing Framework v4.1.1\n
Powered by PHPUnit 9.0.1 by Sebastian Bergmann and contributors.\n
Running with seed:\n
\n
\n
Math2 Tests (1) --------------------------------------------\n
✔ MathCest: Test addition (0.03s)\n
------------------------------------------------------------\n
\n
\n
Time: 384 ms, Memory: 12.00 MB\n
\n
OK (1 test, 2 assertions)\n
\n
\n
Code Coverage Report:\n
  2020-02-29 14:28:26\n
\n
 Summary:\n
  Classes: 50.00% (1/2)\n
  Methods: 50.00% (1/2)\n
  Lines:   25.00% (1/4)\n
\n
\Math2::Math2\Adder\n
  Methods: 100.00% ( 1/ 1)   Lines: 100.00% (  1/  1)\n
Remote CodeCoverage reports are not printed to console\n
' contains "Classes: 100.00%".

Scenario Steps:

 8. $I->seeInShellOutput("Classes: 100.00%") at tests/coverage/LocalShowUncoveredCept.php:13
 7. $I->seeInShellOutput("OK (1 test") at tests/coverage/LocalShowUncoveredCept.php:12
 6. $I->executeCommand("run math2 ...",null,"-d pcov.di...") at tests/coverage/LocalShowUncoveredCept.php:11
 5. $I->seeInShellOutput("Methods: 50.00%") at tests/coverage/LocalShowUncoveredCept.php:8
 4. $I->seeInShellOutput("Classes: 50.00%") at tests/coverage/LocalShowUncoveredCept.php:7
 3. $I->seeInShellOutput("OK (1 test") at tests/coverage/LocalShowUncoveredCept.php:6


FAILURES!
Tests: 1, Assertions: 5, Failures: 1.
```

**Comment by Naktibalda** [comment]
Some relevant improvements were implemented in 9.2.0 version of ph-code-coverage library.
Can we close this issue now?
